library/core/src/iter/adapters/map_windows.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    fmt,
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+/// An iterator over the mapped windows of another iterator.
+///
+/// This `struct` is created by the [`Iterator::map_windows`]. See its
+/// documentation for more information.
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "iter_map_windows", reason = "recently added", issue = "87155")]
+pub struct MapWindows<I: Iterator, F, const N: usize> {
+    iter: I,
+    f: F,
+
+    // The buffer is semantically `[MaybeUninit<I::Item>; 2 * N]`. However, due
+    // to limitations of const generics, we use this different type. Note that
+    // it has the same underlying memory layout.
+    //
+    // Invariant: if `buffer` is `Some`, `buffer[self.start..self.start + N]` is
+    // initialized, with all other elements being uninitialized. This also
+    // implies that `start <= N`.
+    buffer: Option<[[MaybeUninit<I::Item>; N]; 2]>,
+    start: usize,
+}
+
+impl<I: Iterator, F, const N: usize> MapWindows<I, F, N> {
+    pub(in crate::iter) fn new(mut iter: I, f: F) -> Self {
+        assert!(N > 0, "array in `Iterator::map_windows` must contain more than 0 elements");
+
+        let buffer = crate::array::collect_into_array(&mut iter).map(|first_half: [_; N]| {
```

Comment:
Should this advance the underlying iterator? Normally operations like this are deferred to the first call to `Iterator::next` because iterators should be lazy